src/core/common/string.hpp

```diff
@@ -90,76 +90,64 @@ const char *StringFind(const char *aString, char aChar);
  *
  */
-class StringBase
+template <uint16_t kSize> class String
 {
-protected:
+    friend class StringWriter;
+
+    static_assert(kSize > 0, "String buffer cannot be empty.");
+
+public:
     /**
-     * This method appends `printf()` style formatted data to a given character string buffer.
+     * This method returns the string as a null-terminated C string.
      *
-     * @param[in]    aBuffer  A pointer to buffer containing the string.
-     * @param[in]    aSize    The size of the buffer (in bytes).
-     * @param[inout] aLength  A reference to variable containing current length of string. On exit length is updated.
-     * @param[in]    aFormat  A pointer to the format string.
-     * @param[in]    aArgs    Arguments for the format specification.
+     * @returns The null-terminated C string.
      *
-     * @retval kErrorNone          Updated the string successfully.
-     * @retval kErrorNoBufs        String could not fit in the storage.
-     * @retval kErrorInvalidArgs   Arguments do not match the format string.
      */
-    static Error Write(char *aBuffer, uint16_t aSize, uint16_t &aLength, const char *aFormat, va_list aArgs);
+    const char *AsCString(void) const { return mBuffer; }
+
+private:
+    char mBuffer[kSize];
 };
 
-/**
- * This class defines a fixed-size string.
- *
- */
-template <uint16_t SIZE> class String : private StringBase
+class StringWriter
 {
 public:
-    enum
-    {
-        kSize = SIZE, ///< Size (number of characters) in the buffer storage for the string.
-    };
-
     /**
      * This constructor initializes the `String` object as empty.
      *
      */
-    String(void)
-        : mLength(0)
+    StringWriter(char *aBuffer, uint16_t aSize);
+
+    template <uint16_t kSize>
+    explicit StringWriter(String<kSize> &aStringBuffer)
+        : StringWriter(aStringBuffer.mBuffer, kSize)
     {
-        mBuffer[0] = 0;
     }
 
     /**
-     * This constructor initializes the `String` object using `printf()` style formatted data.
+     * This method clears the string.
      *
-     * @param[in] aFormat    A pointer to the format string.
-     * @param[in] ...        Arguments for the format specification.
+     * @returns The string writer.
      *
      */
-    explicit String(const char *aFormat, ...)
-        : mLength(0)
-    {
-        va_list args;
-        va_start(args, aFormat);
-        IgnoreError(Write(mBuffer, kSize, mLength, aFormat, args));
-        va_end(args);
-    }
+    StringWriter &Clear(void);
 
     /**
-     * This method clears the string.
+     * This method returns whether the buffer is sufficient.
+     *
+     * @note If the buffer is insufficient, the output is truncated and null-terminated.
+     *
+     * @retval  true    Buffer is sufficient.
+     * @retval  false   Buffer is insufficient.
      *
      */
-    void Clear(void)
-    {
-        mBuffer[0] = 0;
-        mLength    = 0;
-    }
+    bool IsSufficient(void) const { return mLength < mSize; }
```

Comment:
nit: a user of this method may find something like `IsTruncated()` more intuitive?